engine-upgrade-utils/src/lib.rs

```diff
@@ -0,0 +1,165 @@
+use std::{
+	ffi::{c_void, CString},
+	mem::size_of,
+};
+
+pub mod build_helpers;
+
+// !!!!!!! These constants are used to check the versions across several crates and build scripts.
+// These should be the first things changed when bumping the version, as it will check the
+// rest of the places the version needs changing on build using the build scripts in each of the
+// relevant crates.
+// Should also check that the compatibility function below `args_compatible_with_old` is correct.
+pub const OLD_VERSION: &str = "1.3.3";
+pub const NEW_VERSION: &str = "1.4.0";
+
+pub const ENGINE_LIB_PREFIX: &str = "chainflip_engine_v";
+
+// Sometimes we need to remove arguments that are valid for the new version but not for the old
+// version.
+// The args that are required for 1.4 but *not* 1.3 are:
+// #[derive(Parser, Debug, Clone, Default)]
+// pub struct ArbOptions {
+// 	#[clap(long = "arb.rpc.ws_endpoint")]
+// 	pub arb_ws_endpoint: Option<String>,
+// 	#[clap(long = "arb.rpc.http_endpoint")]
+// 	pub arb_http_endpoint: Option<String>,
+// 	#[clap(long = "arb.private_key_file")]
+// 	pub arb_private_key_file: Option<PathBuf>,
+// }
+pub fn args_compatible_with_old(args: Vec<String>) -> Vec<String> {
+	args.into_iter().filter(|arg| !arg.starts_with("--arb.")).collect()
+}
+
+pub use std::ffi::c_char;
+pub const NO_START_FROM: u32 = 0;
+
+// ====  Status codes ====
+pub const SUCCESS: i32 = 0;
+pub const PANIC: i32 = -1;
+pub const UNKNOWN_ERROR: i32 = -2;
+pub const ERROR_READING_SETTINGS: i32 = -3;
+/// The version of the engine is no longer compatible with the runtime.
+pub const NO_LONGER_COMPATIBLE: i32 = 1;
+/// The engine is not yet compatible with the runtime.
+pub const NOT_YET_COMPATIBLE: i32 = 2;
+
+#[repr(C)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct ExitStatus {
+	pub status_code: i32,
+	pub at_block: u32,
+}
+
+#[link(name = "c")]
+extern "C" {
+	fn malloc(size: usize) -> *mut c_void;
+	fn free(ptr: *mut c_void);
+}
+
+#[repr(C)]
+pub struct CStrArray {
+	// Null pointer if the array isn't initialised.
+	c_args: *mut *mut c_char,
+	n_args: usize,
+}
+
+impl Clone for CStrArray {
+	fn clone(&self) -> Self {
+		let strings = self.to_rust_strings();
+		strings.clone().try_into().unwrap()
+	}
+}
+
+impl Default for CStrArray {
+	fn default() -> Self {
+		Self { c_args: std::ptr::null_mut(), n_args: 0 }
+	}
+}
+
+fn malloc_size<T: Sized>(number_of_ts: usize) -> *mut c_void {
+	unsafe { malloc(size_of::<T>() * number_of_ts) }
+}
+
+impl TryFrom<Vec<String>> for CStrArray {
+	type Error = anyhow::Error;
```

Comment:
Need to check this

Reply:
(I tried it on a localnet, seems to work?)

Reply:
> Need to check this

using anyhow error is fine here as it's only used in the runner (not across dylib boundaries).

> Given that this has caused a few headaches - do we really need it? We only use it to pass in the command line args. Why can the engine lib not pull the args from the environment?

As discussed, to do this we'd need to have a "catch all" settings on each. This isn't ideal because someone can do something `--arp.ws_endpoint` and typo, in order to override their settings file for example, it won't error, but it won't use the specified endpoint because there's a typo. Better to have strict enforcement, and use the explicit filter command `args_compatible_with_old` to do the conversion between the versions.